> In most places I've seen, the placeholder template class is called Tripoint (uppercase) or Point, so I think it makes sense to follow that convention unless there's a reason not to.
> 
> Also note that this only makes sense for relative points, and I think "north" etc. is defined only for those.
> 
> I know you can put constraints on which types to accept, and I think it would be good to do so here to constrain the valid matches to only relative types. If nothing else, it servers as very cryptic documentation. I don't think it would make sense to restrain the template to tripoints, as it would work perfectly fine with (relative) points. I think I've seen Tripoint being used without any dimension constraints to (I assume) serve as an indication of the intended usage, so here I think usage of Tripoint or Point would be a matter of taste (again, if there's no specific reason for the use of "tripo").
> 
> General comment outside the scope of the review. As an aside, yes, I very much prefer having user facing "real" operations with the templates hidden in the implementation. From my perspective, it serves a number of purposes:
> 
> * It tells the user what operations are available, rather than hard to decode templates, typically bunched up among other templates.
> * It tells the compiler what's available, so it won't do stupid things like getting confused when someone does something wrong that matches more than one template, resulting at best in horrible compiler complaint barfs that have to be deciphered, or, at worst, linker errors with no clues at all.
> * It tells the compiler what's available, so it can cleanly tell the user that there's no match (at least VS' Intellisense is template incompetent, so it frequently (always?) won't mark mismatches).

I renamed the point to `CoordPoint` because `Tripoint` no longer makes sense, as we accept `point` too.

I made the template restriction, it took me a while as I am new to it in C++.

I would love to change this:
```cpp
template<typename Point, coords::scale Scale>
static std::optional<coords::coord_point<Point, coords::origin::relative, Scale>>
        get_direction( const std::string &action, bool iso_mode )
{
    using CoordPoint = coords::coord_point<Point, coords::origin::relative, Scale>;
    // function body
```

To something like this (example):
```cpp
template<typename Point, coords::scale Scale>
using CoordPoint = coords::coord_point<Point, coords::origin::relative, Scale>;
static std::optional<CoordPoint>
        get_direction( const std::string &action, bool iso_mode )
{
    // function body
```

But I don't know how. The example code throws `error C3203: 'CoordPoint': unspecialized alias template can't be used as a template argument for template parameter '_Ty', expected a real type`.

--- a/src/input_context.cpp
+++ b/src/input_context.cpp
@@ -548,74 +548,53 @@ static void rotate_direction_cw( int &dx, int &dy )
     dy = dir_num / 3 - 1;
 }
 
-std::optional<tripoint_rel_ms> input_context::get_direction_rel_ms( const std::string &action )
-const
+// This templating ensures that only coord_point with origin::relative is accepted.
+// See src/coords_fwd.h and src/coordinates.h
+template<typename Point, coords::scale Scale>
+static std::optional<coords::coord_point<Point, coords::origin::relative, Scale>>
+        get_direction( const std::string &action, bool iso_mode )
 {
-    static const auto noop = static_cast<tripoint_rel_ms( * )( tripoint_rel_ms )>( [](
-    tripoint_rel_ms p ) {
+    using CoordPoint = coords::coord_point<Point, coords::origin::relative, Scale>;
+    static const auto noop = static_cast<CoordPoint( * )( CoordPoint )>( []( CoordPoint p ) {
         return p;
     } );
-    static const auto rotate = static_cast<tripoint_rel_ms( * )( tripoint_rel_ms )>( [](
-    tripoint_rel_ms p ) {
+    static const auto rotate = static_cast<CoordPoint( * )( CoordPoint )>( []( CoordPoint p ) {
         rotate_direction_cw( p.x(), p.y() );
         return p;
     } );
     const auto transform = iso_mode && g->is_tileset_isometric() ? rotate : noop;
 
     if( action == "UP" ) {
-        return transform( tripoint_rel_ms::north );
+        return transform( CoordPoint::north );
     } else if( action == "DOWN" ) {
-        return transform( tripoint_rel_ms::south );
+        return transform( CoordPoint::south );
     } else if( action == "LEFT" ) {
-        return transform( tripoint_rel_ms::west );
+        return transform( CoordPoint::west );
     } else if( action == "RIGHT" ) {
-        return transform( tripoint_rel_ms::east );
+        return transform( CoordPoint::east );
     } else if( action == "LEFTUP" ) {
-        return transform( tripoint_rel_ms::north_west );
+        return transform( CoordPoint::north_west );
     } else if( action == "RIGHTUP" ) {
-        return transform( tripoint_rel_ms::north_east );
+        return transform( CoordPoint::north_east );
     } else if( action == "LEFTDOWN" ) {
-        return transform( tripoint_rel_ms::south_west );
+        return transform( CoordPoint::south_west );
     } else if( action == "RIGHTDOWN" ) {
-        return transform( tripoint_rel_ms::south_east );
+        return transform( CoordPoint::south_east );
     } else {
         return std::nullopt;
     }
 }
 
+std::optional<tripoint_rel_ms> input_context::get_direction_rel_ms( const std::string &action )
+const
+{
+    return get_direction<tripoint, coords::ms>( action, iso_mode );
+}
+
 std::optional<tripoint_rel_omt> input_context::get_direction_rel_omt( const std::string &action )
 const
 {
-    static const auto noop = static_cast<tripoint_rel_omt( * )( tripoint_rel_omt )>( [](
-    tripoint_rel_omt p ) {
-        return p;
-    } );
-    static const auto rotate = static_cast<tripoint_rel_omt( * )( tripoint_rel_omt )>( [](
-    tripoint_rel_omt p ) {
-        rotate_direction_cw( p.x(), p.y() );
-        return p;
-    } );
-    const auto transform = iso_mode && g->is_tileset_isometric() ? rotate : noop;
-
-    if( action == "UP" ) {
-        return transform( tripoint_rel_omt::north );
-    } else if( action == "DOWN" ) {
-        return transform( tripoint_rel_omt::south );
-    } else if( action == "LEFT" ) {
-        return transform( tripoint_rel_omt::west );
-    } else if( action == "RIGHT" ) {
-        return transform( tripoint_rel_omt::east );
-    } else if( action == "LEFTUP" ) {
-        return transform( tripoint_rel_omt::north_west );
-    } else if( action == "RIGHTUP" ) {
-        return transform( tripoint_rel_omt::north_east );
-    } else if( action == "LEFTDOWN" ) {
-        return transform( tripoint_rel_omt::south_west );
-    } else if( action == "RIGHTDOWN" ) {
-        return transform( tripoint_rel_omt::south_east );
-    } else {
-        return std::nullopt;
-    }
+    return get_direction<tripoint, coords::omt>( action, iso_mode );
 }
 
 // Custom set of hotkeys that explicitly don't include the hardcoded

